Mutable attrs bypass _json_string cache, causing stale hashes and exports

marcorudolphflex opened the pull request:
Problem:
_json_string is cached (tidy3d/components/base.py), but attrs stays mutable. If a model is hashed or serialized before mutating obj.attrs, the cached JSON is reused, so _hash_self() ignores the change and to_hdf5/JSON writes omit the new attrs.

Fixed:

_json_string is cached, but refreshed when attrs refresh

<!-- greptile_comment -->

<h2>Greptile Overview</h2>

Updated On: 2025-10-02 09:22:33 UTC

<h3>Summary</h3>
This PR addresses a critical caching and hashing bug in `Tidy3dBaseModel` related to mutable `attrs` fields. The core issue was that the `_json_string` property was cached without considering changes to the mutable `attrs` dictionary, leading to stale cached values that could cause incorrect hashes and exports.

The solution introduces a sophisticated dual-caching mechanism:

1. **Separated hashing and JSON concerns**: A new `_json_string_for_hashing` property excludes `attrs` entirely to ensure hash stability, while the regular `_json_string` property includes `attrs` with proper cache invalidation.

2. **Smart cache invalidation**: A new `cached_property_guarded` decorator monitors changes to `attrs` through an `_attrs_digest()` method that creates stable fingerprints of the attrs field using JSON encoding and SHA256 hashing.

3. **Consistent export behavior**: The `to_hdf5()` method now uses the appropriate JSON string variant based on whether it's called for hashing purposes.

This change preserves the intended mutability of `attrs` (allowing `obj.attrs['foo'] = bar`) while ensuring that object hashes remain consistent for objects with the same core properties, regardless of metadata changes. The JSON serialization still properly reflects the current state including any attrs modifications through the guarded caching mechanism.

## Important Files Changed

<details>
<summary>Changed Files</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| `tidy3d/components/base.py` | 4/5 | Implements sophisticated dual-caching mechanism with guarded properties to fix mutable attrs cache invalidation issues |
| `tests/test_components/test_base.py` | 5/5 | Adds comprehensive tests validating hash stability and JSON cache invalidation for mutable attrs |
| `CHANGELOG.md` | 4/5 | Documents the bug fix, though categorized under 'Changed' rather than 'Fixed' section |

</details>

## Confidence score: 4/5

- This PR addresses a well-defined bug with a thoughtful solution that maintains backward compatibility while fixing caching issues
- Score reflects the complexity of the caching mechanism which requires careful review to ensure no edge cases are missed
- The `CHANGELOG.md` categorization should be reviewed as this appears to be a bug fix rather than a feature change

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant BaseModel
    participant Cache
    participant HashFunction
    
    User->>BaseModel: Create object with attrs
    BaseModel->>Cache: Store initial _json_string
    BaseModel->>HashFunction: Generate _hash_self (excludes attrs)
    
    User->>BaseModel: Modify attrs["foo"] = "new_value"
    BaseModel->>BaseModel: Call _attrs_digest()
    BaseModel->>Cache: Invalidate _json_string cache (key changed)
    BaseModel->>Cache: Generate new _json_string (includes updated attrs)
    
    User->>BaseModel: Call _hash_self()
    BaseModel->>HashFunction: Return same hash (attrs excluded)
    Note over BaseModel,HashFunction: Hash remains stable despite attrs change
    
    User->>BaseModel: Access _json_string
    BaseModel->>Cache: Return updated cached value
    Note over BaseModel,Cache: Cache reflects current attrs state
```

<!-- greptile_other_comments_section -->

**Context used:**

Rule from `dashboard` - Use changelog categories correctly: "Fixed" for bug fixes, "Changed" for modifications to existing f... ([source](https://app.greptile.com/review/custom-context?memory=17997812-233f-4a81-bee2-4d7666605061))

<!-- /greptile_comment -->

yaugenst-flex commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 - Improved performance of antenna metrics calculation by utilizing cached wave amplitude calculations instead of recomputing wave amplitudes for each port excitation in the `TerminalComponentModelerData`.
+- Fixed: In `Tidy3dBaseModel` the hash (and cached `.json_string`) are now sensitive to changes in `.attrs`.
+- Changed hashing method in `Tidy3dBaseModel` from sha256 to md5.
```
The fix should go under `Fixed`

yaugenst-flex reviewed (changes requested):
Thanks @marcorudolphflex looks great! Just a minor nit on the changelog.

Oh and the commit should be tagged `fix:`, not `feat:`

marcorudolphflex committed:
fix(tidy3d): Mutable attrs bypass _json_string cache, causing stale...
hashes and exports

- _json_string is cached, but refreshed when attrs refresh
- changed hashing method from sha256 to md5

Jira: FXC-3374
